File: 2018/07-jul/7-29-overload-override/java/Garden.java
import java.util.ArrayList;
import java.util.List;

/**
 * Class representing a garden that I do work in at Tod's Point.  The class is used to demonstrate the
 * differences between overloading and overriding.  It is also immutable.
 * @author Andrew Jarombek
 * @since 7/21/2018
 */
public final class Garden {

    private List<Animal> animals;
    private List<Plant> plants;
    private List<Object> randomObjects;

    /**
     * Private constructor to disallow subclassing and enforce the use of static factory methods
     * @param animals a group of animals, the class must implement {@code Iterable}
     * @param plants a group of plants, the class must implement {@code Iterable}
     * @param randomObjects a group of random objects, the class must implement {@code Iterable}
     */
    private Garden(Iterable<Animal> animals, Iterable<Plant> plants, Iterable<?> randomObjects) {
        this.animals = new ArrayList<>();
        this.plants = new ArrayList<>();
        this.randomObjects = new ArrayList<>();

        animals.forEach(this.animals::add);
        plants.forEach(this.plants::add);
        randomObjects.forEach(this.randomObjects::add);
    }

    /**
     * Static factory method to create a Garden.  No random objects will exist in the garden.
     * @param animals a group of animals, the class must implement {@code Iterable}
     * @param plants a group of plants, the class must implement {@code Iterable}
     * @return a new Garden instance
     */
    static Garden of(Iterable<Animal> animals, Iterable<Plant> plants) {
        return new Garden(animals, plants, List.of());
    }

    /**
     * Static factory method to create a Garden.  No random objects will exist in the garden.
     * @param animals a group of animals, the class must implement {@code Iterable}
     * @param plants a group of plants, the class must implement {@code Iterable}
     * @param randomObjects a group of random objects, the class must implement {@code Iterable}
     * @return a new Garden instance
     */
    static Garden of(Iterable<Animal> animals, Iterable<Plant> plants, Iterable<?> randomObjects) {
        return new Garden(animals, plants, randomObjects);
    }

    /**
     * Check if the garden instance contains the provided animal
     * @param animal an animal to search the internal list structure for
     * @return {@code true} if the garden contains the animal, {@code false} otherwise
     */
    boolean inGarden(Animal animal) {
        System.out.println("Checking if Animal in Garden");
        return animals.contains(animal);
    }

    /**
     * Check if the garden instance contains the provided plant
     * @param plant a plant to search the internal list structure for
     * @return {@code true} if the garden contains the plant, {@code false} otherwise
     */
    boolean inGarden(Plant plant) {
        System.out.println("Checking if Plant in Garden");
        return plants.contains(plant);
    }

    /**
     * Check if the garden instance contains the provided object
     * @param object a random object to search the internal list structure for
     * @return {@code true} if the garden contains the object, {@code false} otherwise
     */
    boolean inGarden(Object object) {
        System.out.println("Checking if Random Object in Garden");
        return randomObjects.contains(object);
    }
}

File: 2018/07-jul/7-29-overload-override/javascript/Garden.js
/**
 * Representation of a Garden I do work in at Tod's Point
 * @author Andrew Jarombek
 * @since 7/27/2018
 */
class Garden {

    /**
     * Constructor for a Garden containing different animals, plants, and other objects.  All the
     * arguments for the constructor are optional and will be initialized as an empty array if
     * they are missing
     * @param animals - a list of animals in the garden (defaults to an empty array)
     * @param plants - a list of plants in the garden (defaults to an empty array)
     * @param randomObjects - a list of random objects in the garden (defaults to an empty array)
     */
    constructor(animals=[], plants=[], randomObjects=[]) {
        this.animals = animals;
        this.plants = plants;
        this.randomObjects = randomObjects;
    }

    /**
     * Check if an item exists in the garden.  If the argument passed in is non existent or not an
     * object, {false} is returned.  Through duck typing checks, if the object passed is an animal,
     * the internal {animals} array is checked.  If the object is a plant, the array {plants} is
     * checked.  Otherwise, the array {randomObjects} is checked.
     * @param item - an item that can be an animal, plant, or any random object
     * @return {boolean} whether or not the item exists in the garden
     */
    inGarden(item) {
        if (arguments.length === 0) {
            return false;
        }

        if (typeof item === "object") {

            // Check for the object type using duck typing
            if (item.hasOwnProperty("name") && item.hasOwnProperty("species")
                    && item.hasOwnProperty("description")) {
                return this.animals.includes(item);
            } else if (item.hasOwnProperty("species") && item.hasOwnProperty("inBloom")) {
                return this.plants.includes(item);
            } else {
                return this.randomObjects.includes(item);
            }

        } else {
            return false;
        }
    }
}

module.exports = Garden;